temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java

```diff
@@ -228,16 +247,60 @@ private void handleEventsBatch(List<HistoryEvent> events, boolean hasNextEvent)
       }
     }
 
+    // Look ahead to infer protocol messages
+    if (EventType.EVENT_TYPE_WORKFLOW_TASK_STARTED.equals(events.get(0).getEventType())) {
+      if (this.isReplaying()) {
+        for (HistoryEvent event : events) {
+          if (event.hasWorkflowExecutionUpdateAcceptedEventAttributes()) {
+            WorkflowExecutionUpdateAcceptedEventAttributes updateEvent =
+                event.getWorkflowExecutionUpdateAcceptedEventAttributes();
+            this.messages.add(
+                Message.newBuilder()
+                    .setId(updateEvent.getAcceptedRequestMessageId())
+                    .setProtocolInstanceId(updateEvent.getProtocolInstanceId())
+                    .setEventId(updateEvent.getAcceptedRequestSequencingEventId())
+                    .setBody(Any.pack(updateEvent.getAcceptedRequest()))
+                    .build());
+          }
+        }
+      }
+    }
+
     for (Iterator<HistoryEvent> iterator = events.iterator(); iterator.hasNext(); ) {
       HistoryEvent event = iterator.next();
+
+      // On replay the messages are available after the workflow task schedule event, so we
+      // need to handle them before workflow task started event to maintain a consistent order.
+      for (Message msg : this.takeLTE(event.getEventId() - 1)) {
```

Comment:
This is not very elegant, but It tried a few ways and this is the best I could come up with that doesn't require a lot of refactoring. Keep in mine once we have proper commands for messages this code will need to change so I didn't think it was worth the effort at this moment.

Reply:
This hack was necessary because of a quirk in how Go handles events, namely that WFTScheduled events don't even go through the normal event handler with the big event type switch. If Java doesn't have that same quirk then you don't need to replicate that weirdness.